lib/Arch/Runtime/HyperCall.cpp

```diff
@@ -0,0 +1,188 @@
+/*
+ * Copyright (c) 2022 Trail of Bits, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+Memory *__remill_sync_hyper_call(State &state, Memory *mem,
+                                 SyncHyperCall::Name call) {
+  switch (call) {
+
+#if defined(REMILL_ON_X86) or defined(REMILL_ON_AMD64)
+
+    case SyncHyperCall::kX86CPUID:
+      asm volatile("cpuid"
+                   : "=a"(state.gpr.rax.aword), "=b"(state.gpr.rbx.aword),
+                     "=c"(state.gpr.rcx.aword), "=d"(state.gpr.rdx.aword)
+                   : "a"(state.gpr.rax.aword), "b"(state.gpr.rbx.aword),
+                     "c"(state.gpr.rcx.aword), "d"(state.gpr.rdx.aword));
+      break;
+
+    case SyncHyperCall::kX86ReadTSC:
+      asm volatile("rdtsc"
+                   : "=a"(state.gpr.rax.dword), "=d"(state.gpr.rdx.dword));
+      break;
+
+    case SyncHyperCall::kX86ReadTSCP:
+      asm volatile("rdtscp"
+                   : "=a"(state.gpr.rax.aword), "=c"(state.gpr.rcx.aword),
+                     "=d"(state.gpr.rdx.aword)
+                   : "a"(state.gpr.rax.aword), "c"(state.gpr.rcx.aword),
+                     "d"(state.gpr.rdx.aword));
+      break;
+
+    case SyncHyperCall::kX86LoadGlobalDescriptorTable:
+      // TODO(alex): Where do I get the operand from?
+      asm volatile("lgdt" : : "=m"(__remill_read_memory_64(memory, 0)));
+      break;
+
+    case SyncHyperCall::kX86LoadInterruptDescriptorTable:
+      // TODO(alex): Where do I get the operand from?
+      asm volatile("lidt" : : "m"(__remill_read_memory_64(memory, 0)));
+      break;
+
+    case SyncHyperCall::kX86ReadModelSpecificRegister:
+      asm volatile("rdmsr"
+                   : "=c"(state.gpr.rcx.dword)
+                   : "a"(state.gpr.rax.dword), "d"(state.gpr.rdx.dword));
+      break;
+
+    case SyncHyperCall::kX86WriteModelSpecificRegister:
+      asm volatile("wrmsr"
+                   : "=c"(state.gpr.rcx.dword)
+                   : "a"(state.gpr.rax.dword), "d"(state.gpr.rdx.dword));
+      break;
+
+    case SyncHyperCall::kX86WriteBackInvalidate:
+      // NOTE(alex): This just clears cache so there's no affect on the state.
+      asm volatile("wbinvd" :);
+      break;
+
+    // TODO(alex): There doesn't seem to be a way to figure out what what value
```

Comment:
@pgoodman Do you know the reason behind this?

Reply:
I am not sure. I'd have to look at the arch docs to see how this instruction works.

Reply:
This one doesn't map to an instruction. It just gets as part of other instructions that eventually write to a segment register. What I don't understand is that the `State` structure has these segment registers.

Here's an instruction where it uses the hyper call:
https://github.com/lifting-bits/remill/blob/master/lib/Arch/X86/Semantics/POP.cpp#L163-L171

And here are the segment registers:
https://github.com/lifting-bits/remill/blob/master/include/remill/Arch/X86/Runtime/State.h#L463-L476

There are also some other registers called `es_base`, `ss_base`, etc. These ones actually get written to in some of the semantic functions:
https://github.com/lifting-bits/remill/blob/master/include/remill/Arch/X86/Runtime/State.h#L672-L685
https://github.com/lifting-bits/remill/blob/master/lib/Arch/X86/Semantics/CALL_RET.cpp#L26

I'm not sure what the distinction between the base and non-base variants is because the segment registers are also meant to hold the "base" address of the segment.

Reply:
From my memory, base address of the segment is "hidden." For practical purposes, we need to know it, though, as that hidden base is used in calculations what involve implicit or explicit segment register use.

```c++
WritePtr<addr_t>(next_sp _IF_32BIT(REG_SS_BASE))
```

That `WritePtr` is sort of a fancy cast, and conditional use of `ss_base` is saying "maybe displace by `ss_base`, i.e. `next_sp + ss_base`, so it's actually read in this instance.

I think I'd need to see the instructions that you're working with to put the picture together and contextualize this a bit more.

Reply:
> I think I'd need to see the instructions that you're working with to put the picture together and contextualize this a bit more.

I'm not seeing these lifted in an example, I'm just going through each hyper call type and trying to implement it.